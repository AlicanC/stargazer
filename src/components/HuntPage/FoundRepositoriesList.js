// @flow

import * as React from 'react';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { ListGroup, ListGroupItem } from 'reactstrap';

import RepositoryView from '../RepositoryView';

const FoundRepositoriesQuery = gql`
  query FoundRepositoriesQuery($query: String!, $cursor: String) {
    search(type: REPOSITORY, first: 10, query: $query, after: $cursor) {
      edges {
        node {
          ... on Repository {
            id
            ...RepositoryViewRepository
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
  ${RepositoryView.fragments.repository}
`;

type Props = {
  data: any,
};

function FoundRepositoriesList(props: Props) {
  const { data } = props;

  if (data.loading) return 'Loading...';
  if (data.error) return 'Error: ' + data.error.message;

  const { search } = data;
  const { hasNextPage } = search.pageInfo;

  const fetchMore = () => {
    data.fetchMore({
      variables: {
        cursor: search.pageInfo.endCursor,
      },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        const newEdges = fetchMoreResult.search.edges;
        const pageInfo = fetchMoreResult.search.pageInfo;

        return newEdges.length
          ? {
              search: {
                __typename: previousResult.search.__typename,
                edges: [...previousResult.search.edges, ...newEdges],
                pageInfo,
              },
            }
          : previousResult;
      },
    });
  };

  return (
    <>
      <ListGroup>
        {search.edges.map(({ node: repository }) => (
          <ListGroupItem key={repository.id}>
            <RepositoryView repository={repository} />
          </ListGroupItem>
        ))}
      </ListGroup>
      {hasNextPage && (
        <div className="my-3" style={{ textAlign: 'right' }}>
          <button onClick={fetchMore}>See More</button>
        </div>
      )}
    </>
  );
}

export default graphql(FoundRepositoriesQuery, {
  options: (props) => ({
    variables: {
      query: props.query,
    },
  }),
})(FoundRepositoriesList);
