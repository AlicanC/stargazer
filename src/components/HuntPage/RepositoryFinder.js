// @flow

import * as React from 'react';

import FoundRepositoriesList from './FoundRepositoriesList';

type State = {
  query: string,
};

export default class RepositoryFinder extends React.Component<{}, State> {
  state = {
    query: '',
  };

  onQueryInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const {
      target: { value: query },
    } = event;

    this.setState({ query });
  };

  render() {
    const { query } = this.state;

    return (
      <div>
        <input type="search" onChange={this.onQueryInputChange} />
        <FoundRepositoriesList query={query} />
      </div>
    );
  }
}
