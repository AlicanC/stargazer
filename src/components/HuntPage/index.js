// @flow

import * as React from 'react';
import { Container, Col, Row, Input } from 'reactstrap';
import { connect } from 'react-redux';

import LoginRequiredView from '../LoginRequiredView';

import FoundRepositoriesList from './FoundRepositoriesList';

type Props = {
  token: ?string,
};

type State = {
  query: string,
};

class HuntPage extends React.PureComponent<Props, State> {
  state = {
    query: '',
  };

  onQueryInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const {
      target: { value: query },
    } = event;

    this.setState({ query });
  };

  render() {
    const { token } = this.props;
    const { query } = this.state;

    return (
      <Container>
        {token ? (
          <>
            <Row className="mb-3">
              <Col style={{ textAlign: 'center' }}>
                <h5>Get ready to hunt!</h5>
                Search for superstar repos and star them!
              </Col>
            </Row>
            <Row className="mb-3">
              <Col>
                <Input type="search" onChange={this.onQueryInputChange} />
              </Col>
            </Row>
            {query.length ? (
              <Row>
                <Col>
                  <FoundRepositoriesList query={query} />
                </Col>
              </Row>
            ) : null}
          </>
        ) : (
          <LoginRequiredView />
        )}
      </Container>
    );
  }
}

export default connect((state) => ({
  token: state.token,
}))(HuntPage);
