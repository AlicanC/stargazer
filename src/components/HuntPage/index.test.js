import React from 'react';
import TestRenderer from 'react-test-renderer';

import HuntPage from '.';

jest.mock('react-router-dom', () => ({
  Link: 'div',
}));

jest.mock('react-redux', () => ({
  connect: () => (C) => C,
}));

it('matches snapshot', () => {
  const renderer = TestRenderer.create(<HuntPage />);

  expect(renderer.toJSON()).toMatchSnapshot();
});
