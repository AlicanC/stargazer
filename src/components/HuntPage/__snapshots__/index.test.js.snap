// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="text-center"
  >
    <h5>
      Please login to view this page.
    </h5>
    <div
      to="/login"
    >
      Login
    </div>
  </div>
</div>
`;
